### PR TITLE
Anuragd/dla support l4t

### PR DESCRIPTION
**Pull request by andi4191**
# Description
Changes include:

Device selection API introduced (For multi-GPU systems)
DLA support for NVIDIA Jetson AGX platform
Test script for API validation on DLA in half-precision
An abstraction of device selection based on TensorRT engine [Removed from this PR]
Fixes # (issue)

Type of change
Breaking change

Breaking change (Python binding modified for DeviceType)

This change requires a documentation update

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/accuracy/test_fp32_accuracy.cpp [hunk 1]
```diff
@@ -50,6 +50,6 @@ TEST_P(AccuracyTests, FP16AccuracyIsClose) {
 }
 
 INSTANTIATE_TEST_SUITE_P(
-    FP16AccuracyIsCloseSuite,
+    FP32AccuracyIsCloseSuite,
     AccuracyTests,
-    testing::Values("tests/accuracy/vgg16_cifar10.jit.pt"));
+    testing::Values("tests/accuracy/resnet50_scripted.jit.pt"));
```
Comment (narendasan) [inline tests/accuracy/test_fp32_accuracy.cpp:55]: can we keep this as VGG?

File: tests/accuracy/test_fp16_accuracy.cpp [hunk 2]
```diff
@@ -52,4 +52,4 @@ TEST_P(AccuracyTests, FP16AccuracyIsClose) {
 INSTANTIATE_TEST_SUITE_P(
     FP16AccuracyIsCloseSuite,
     AccuracyTests,
-    testing::Values("tests/accuracy/vgg16_cifar10.jit.pt"));
+    testing::Values("tests/accuracy/resnet50_scripted.jit.pt"));
```
Comment (narendasan) [inline tests/accuracy/test_fp16_accuracy.cpp:55]: can we keep this as VGG?

File: third_party/tensorrt/local/BUILD [hunk 3]
```diff
@@ -67,11 +67,6 @@ cc_import(
         ":windows": "lib/nvinfer.dll",
         "//conditions:default": "lib/x86_64-linux-gnu/libnvinfer.so",
     }),
-    static_library = select({
```
Comment (narendasan) [inline third_party/tensorrt/local/BUILD:70]: Id still like to solve this correctly. Can you take a look at the places where linkstatic is used? I think thats the root cause.

File: tests/accuracy/test_fp32_accuracy.cpp (hunk 1 above)
Comment (andi4191) [reply tests/accuracy/test_fp32_accuracy.cpp:55]: Where can I get vgg16_cifar10.jit.pt ?
I couldn't find it so I used resnet50_scripted.jit.pt from hub.py scripts

File: third_party/tensorrt/local/BUILD (hunk 3 above)
Comment (andi4191) [reply third_party/tensorrt/local/BUILD:70]: It throws undefined references to all the TensorRT symbols on Jetson AGX

Comment (narendasan) [reply third_party/tensorrt/local/BUILD:70]: lets then break out the static library into a different target,

File: tests/accuracy/test_fp16_accuracy.cpp (hunk 2 above)
Comment (andi4191) [reply tests/accuracy/test_fp16_accuracy.cpp]: done

File: tests/accuracy/test_fp32_accuracy.cpp (hunk 1 above)
Comment (andi4191) [reply tests/accuracy/test_fp32_accuracy.cpp]: done

File: third_party/tensorrt/local/BUILD (hunk 3 above)
Comment (andi4191) [reply third_party/tensorrt/local/BUILD:70]: separated the static library target

**Review by narendasan — COMMENTED** [review]
Did you implement the changes needed for to_backend or do i need to do that?

File: cpp/api/include/trtorch/trtorch.h [hunk 4]
```diff
@@ -334,16 +335,42 @@ struct TRTORCH_API CompileSpec {
...
+     * This struct will hold Target device related parameters such as device_type, gpu_id, dla_core
+     */
+    DeviceType device_type;
 
-  /**
-   * Target device type
+    /*
+     * Target gpu id
+     */
+    int64_t gpu_id;
```
[the remark this answers is not in the document]
Comment (andi4191) [reply cpp/api/include/trtorch/trtorch.h:351]: Constructor takes care of it already.

File: cpp/api/include/trtorch/trtorch.h [hunk 5]
```diff
@@ -334,16 +335,42 @@ struct TRTORCH_API CompileSpec {
...
+  /*
+   * Setting data structure for Target device
    */
-  bool allow_gpu_fallback = true;
+  struct Device {
+    /**
+     * @brief Setting data structure for device
+     * This struct will hold Target device related parameters such as device_type, gpu_id, dla_core
+     */
+    DeviceType device_type;
```
[the remark this answers is not in the document]
Comment (andi4191) [reply cpp/api/include/trtorch/trtorch.h:346]: What about the class DeviceType in trtorch.h ?

File: py/trtorch/_compile_spec.py [hunk 6]
```diff
@@ -175,12 +198,16 @@ def TensorRTCompileSpec(compile_spec: Dict[str, Any]):
                                 "max": (1, 3, 1024, 1024)
                             } # Dynamic input shape for input #2
                         ],
+                        "device": {
+                        "device_type": torch.device("cuda"), # Type of device to run engine on (for DLA use trtorch.DeviceType.DLA)
```
[the remark this answers is not in the document]
Comment (andi4191) [reply py/trtorch/_compile_spec.py:202]: It is in the documentation and already tabbed in.

File: py/trtorch/csrc/trtorch_py.cpp [hunk 7]
```diff
@@ -86,8 +86,8 @@ PYBIND11_MODULE(_C, m) {
       .export_values();
 
   py::enum_<DeviceType>(m, "DeviceType", "Enum to specify device kinds to build TensorRT engines for")
-      .value("gpu", DeviceType::kGPU, "Specify using GPU to execute TensorRT Engine")
-      .value("dla", DeviceType::kDLA, "Specify using DLA to execute TensorRT Engine (Jetson Only)")
+      .value("GPU", DeviceType::kGPU, "Specify using GPU to execute TensorRT Engine")
```
[the remark this answers is not in the document]
Comment (andi4191) [reply py/trtorch/csrc/trtorch_py.cpp:89]: Updated the MR details.

File: cpp/api/include/trtorch/trtorch.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (andi4191) [reply cpp/api/include/trtorch/trtorch.h:346]: Fixed it

**Comment by narendasan** [comment]
Can you merge in the changes to the build system for tests that I made?

**Comment by andi4191** [comment]
@narendasan: Pushed in the changes for aarch64_tests for modules

**Review by narendasan — APPROVED** [review]
LGTM to merge

**Pull request merged**